File: php/telco_system/tests/TelcoSystemTest.php
<?php

use Carbon\Carbon;
use Carbon\CarbonImmutable;
use PHPUnit\Framework\TestCase;

class TelcoSystemTest extends TestCase
{
    private $aDateTime;
    private $anotherDateTime;

    protected function setUp(): void
    {
        parent::setUp();

        $this->aDateTime = CarbonImmutable::now();
        $this->anotherDateTime = CarbonImmutable::now();
    }

    /**
     * @test
     */
    public function initialHistoricalTotalBilledShouldBeZero()
    {
        $telcoSystem = new TelcoSystem();
        $this->assertEquals(0, $telcoSystem->historicalTotalBilled());
    }

    /**
     * @test
     */
    public function internationalCallOfTwoMinutesShouldCost4_pesos()
    {
        $telcoSystem = new TelcoSystem();
        $telcoSystem->registerInternationalCallBetween($this->aDateTime, $this->aDateTime->addMinutes(2), "Juan");

        $this->assertEquals(4, $telcoSystem->historicalTotalBilled());
        $this->assertEquals(1, $telcoSystem->historicalNumberOfCalls());
    }

    /**
     * @test
     */
    public function nationalCallOf100MinutesShouldCost150Pesos()
    {
        $telcoSystem = new TelcoSystem();
        $telcoSystem->registerNationalCallBetween($this->aDateTime, $this->aDateTime->addMinutes(100), "Juan");

        $this->assertEquals(150, $telcoSystem->historicalTotalBilled());
        $this->assertEquals(1, $telcoSystem->historicalNumberOfCalls());
    }

    /**
     * @test
     */
    public function localCallOf10MinutesDuringRushHourShouldCost10Pesos()
    {
        $telcoSystem = TelcoSystem::withRushHourStartingAt(18);
        $eighteenO_clock = CarbonImmutable::createFromTime(18, 0, 0, 'Europe/London');

        $telcoSystem->registerLocalCallBetween($eighteenO_clock, $eighteenO_clock->addMinutes(10), "Juan");

        $this->assertEquals(10, $telcoSystem->historicalTotalBilled());
        $this->assertEquals(1, $telcoSystem->historicalNumberOfCalls());
    }

    /**
     * @test
     */
    public function localCallOf6MinutesDuringNotRushHourShouldCost3_pesos()
    {
        $telcoSystem = TelcoSystem::withRushHourStartingAt(18);
        $twoO_clock = CarbonImmutable::createFromTime(14, 0, 0, 'Europe/London');

        $telcoSystem->registerLocalCallBetween($twoO_clock, $twoO_clock->addMinutes(6), "Juan");

        $this->assertEquals(3, $telcoSystem->historicalTotalBilled());
        $this->assertEquals(1, $telcoSystem->historicalNumberOfCalls());
    }

    /**
     * @test
     */
    public function localCallStartedInRushHourAndFinishedInNonRushHourShouldConsiderEachCost()
    {

        $telcoSystem = TelcoSystem::withRushHourStartingAt(18);
        $twoMinutesBeforeRushHour = CarbonImmutable::createFromTime(17, 58, 0, 'Europe/London');

        $telcoSystem->registerLocalCallBetween($twoMinutesBeforeRushHour, $twoMinutesBeforeRushHour->addMinutes(6), "Juan");

        $this->assertEquals(5, $telcoSystem->historicalTotalBilled());
        $this->assertEquals(1, $telcoSystem->historicalNumberOfCalls());
    }

    /**
     * @test
     */
    public function multipleClientsShouldBeSupported()
    {
        $telcoSystem = new TelcoSystem();

        $telcoSystem->registerInternationalCallBetween($this->aDateTime, $this->aDateTime->addMinutes(2), "Juan");
        $telcoSystem->registerInternationalCallBetween($this->anotherDateTime, $this->anotherDateTime->addMinutes(2), "Juan");

        $telcoSystem->registerInternationalCallBetween($this->aDateTime, $this->aDateTime->addMinutes(20), "Pedro");
        $telcoSystem->registerInternationalCallBetween($this->anotherDateTime, $this->anotherDateTime->addMinutes(12), "Pedro");

        $this->assertEquals(8, $telcoSystem->totalBilledFor("Juan"));

        $this->assertEquals(2, $telcoSystem->numberOfCallsFor("Juan"));
        $this->assertEquals(4, $telcoSystem->historicalNumberOfCalls());
    }
}

File: java/telco_system/src/test/java/com/devartis/exercise/telco/TelcoSystemTest.java
package com.devartis.exercise.telco;

import org.joda.time.DateTime;
import org.junit.Test;

import static org.junit.Assert.assertEquals;

public class TelcoSystemTest {

    private DateTime aDateTime = DateTime.now();
    private DateTime anotherDateTime = aDateTime.plusDays(2);
    private double delta = 0.01;

    @Test
    public void theHistoricalTotalBilledShouldBeZeroIfNoCallsWereMade() {
        TelcoSystem telcoSystem = new TelcoSystem();

        assertEquals(0, telcoSystem.historicalTotalBilled(), delta);
    }

    @Test
    public void anInternationalCallOfTwoMinutesShouldCost4pesos() {
        TelcoSystem telcoSystem = new TelcoSystem();

        telcoSystem.registerInternationalCallBetween(aDateTime, aDateTime.plusMinutes(2), "Juan");

        assertEquals(4, telcoSystem.historicalTotalBilled(), delta);
        assertEquals(1, telcoSystem.historicalNumberOfCalls());
    }

    @Test
    public void aNationalCallOf100MinutesShouldCost150pesos() {
        TelcoSystem telcoSystem = new TelcoSystem();
        telcoSystem.registerNationalCallBetween(aDateTime, aDateTime.plusMinutes(100), "Juan");

        assertEquals(150, telcoSystem.historicalTotalBilled(), delta);
        assertEquals(1, telcoSystem.historicalNumberOfCalls());
    }

    @Test
    public void aLocalCallOf10MinutesDuringPeakHourShouldCost10pesos() {
        TelcoSystem telcoSystem = TelcoSystem.withPeakHourStartingAt(18);
        DateTime eighteenOClock = DateTime.now().withHourOfDay(18).withMinuteOfHour(0);

        telcoSystem.registerLocalCallBetween(eighteenOClock, eighteenOClock.plusMinutes(10), "Juan");

        assertEquals(10, telcoSystem.historicalTotalBilled(), delta);
        assertEquals(1, telcoSystem.historicalNumberOfCalls());
    }

    @Test
    public void aLocalCallOf6MinutesDuringNotPeakHourShouldCost3pesos() {
        TelcoSystem telcoSystem = TelcoSystem.withPeakHourStartingAt(18);
        DateTime twoOClock = DateTime.now().withHourOfDay(14).withMinuteOfHour(0);

        telcoSystem.registerLocalCallBetween(twoOClock, twoOClock.plusMinutes(6), "Juan");

        assertEquals(3, telcoSystem.historicalTotalBilled(), delta);
        assertEquals(1, telcoSystem.historicalNumberOfCalls());
    }

    @Test
    public void aLocalCallStartedBeforePeakHourAndFinishedInPeakHourShouldConsiderEachCost() {
        TelcoSystem telcoSystem = TelcoSystem.withPeakHourStartingAt(18);
        DateTime twoMinutesBeforePeakHour = DateTime.now().withHourOfDay(17).withMinuteOfHour(58);

        telcoSystem.registerLocalCallBetween(twoMinutesBeforePeakHour, twoMinutesBeforePeakHour.plusMinutes(6), "Juan");

        assertEquals(5, telcoSystem.historicalTotalBilled(), delta);
        assertEquals(1, telcoSystem.historicalNumberOfCalls());
    }

    @Test
    public void aLocalCallStartedAfterPeakHourAndFinishedInNonPeakHourShouldConsiderEachCost() {
        TelcoSystem telcoSystem = TelcoSystem.withPeakHourStartingAt(18);
        DateTime twoMinutesBeforePickHour = DateTime.now().withHourOfDay(18).withMinuteOfHour(58);

        telcoSystem.registerLocalCallBetween(twoMinutesBeforePickHour, twoMinutesBeforePickHour.plusMinutes(6), "Juan");

        assertEquals(4, telcoSystem.historicalTotalBilled(), delta);
        assertEquals(1, telcoSystem.historicalNumberOfCalls());
    }
    
    @Test
    public void twoInternationalCallsOf2MinutesShouldCost8PesosForJuan() {
       TelcoSystem telcoSystem = new TelcoSystem();

        telcoSystem.registerInternationalCallBetween(aDateTime, aDateTime.plusMinutes(2), "Juan");
        telcoSystem.registerInternationalCallBetween(anotherDateTime, anotherDateTime.plusMinutes(2), "Juan");

        telcoSystem.registerInternationalCallBetween(aDateTime, aDateTime.plusMinutes(20), "Pedro");
        telcoSystem.registerInternationalCallBetween(anotherDateTime, anotherDateTime.plusMinutes(12), "Pedro");

        assertEquals(8, telcoSystem.totalBilledFor("Juan"), delta);

        assertEquals(2, telcoSystem.numberOfCallsFor("Juan"));
        assertEquals(4, telcoSystem.historicalNumberOfCalls());
    }
}